File: drouter-plugin/src/main/groovy/com/didi/drouter/utils/StoreUtil.java
package com.didi.drouter.utils;

import java.util.HashMap;
import java.util.Map;

import javassist.CtClass;

/**
 * Created by gaowei on 2018/10/30
 */
public class StoreUtil {

    private static Map<String, String> pathMap = new HashMap<>();
    private static Map<String, Class<?>> classMap = new HashMap<>();
    private static Map<String, String> callAliasMap = new HashMap<>();

    public static String insertUri(String uri, CtClass cc) {
        if (!pathMap.containsKey(uri)) {
            pathMap.put(uri, cc.getName());
            return null;
        }
        return pathMap.get(uri);
    }

    public static Class<?> getClass(CtClass ctClass, ClassLoader loader) throws ClassNotFoundException {
        if (classMap.containsKey(ctClass.getName())) {
            return classMap.get(ctClass.getName());
        } else {
            //Class<?> clz = ctClass.toClass(loader, null);  // throw duplicate error, no reuse
            Class<?> clz = Class.forName(ctClass.getName(), false, loader);
            ctClass.defrost();
            classMap.put(ctClass.getName(), clz);
            return clz;
        }
    }

    public static void clear() {
        pathMap.clear();
        classMap.clear();
        callAliasMap.clear();
    }

    public static String insertCallAlias(String callAlias, CtClass cc) {
        if (TextUtil.isEmpty(callAlias)) {
            return null;
        }
        if (!callAliasMap.containsKey(callAlias)) {
            callAliasMap.put(callAlias, cc.getName());
            return null;
        }
        return callAliasMap.get(callAlias);
    }
}

File: plugins/drouter-plugin/src/main/kotlin/com/didi/drouter/utils/StoreUtil.kt
package com.didi.drouter.utils

import javassist.CtClass

/**
 * Created by gaowei on 2018/10/30
 */
object StoreUtil {
    private val pathMap: MutableMap<String, String> = HashMap()
    private val classMap: MutableMap<String, Class<*>> = HashMap()
    private val callAliasMap: MutableMap<String, String> = HashMap()
    fun insertUri(uri: String, cc: CtClass): String? {
        if (!pathMap.containsKey(uri)) {
            pathMap[uri] = cc.name
            return null
        }
        return pathMap[uri]
    }

    @Throws(ClassNotFoundException::class)
    fun getClass(ctClass: CtClass, loader: ClassLoader?): Class<*> {
        return if (classMap.containsKey(ctClass.name)) {
            classMap[ctClass.name]!!
        } else {
            //Class<?> clz = ctClass.toClass(loader, null);  // throw duplicate error, no reuse
            val clz = Class.forName(ctClass.name, false, loader)
            classMap[ctClass.name] = clz
            clz
        }
    }

    fun clear() {
        pathMap.clear()
        classMap.clear()
        callAliasMap.clear()
    }

    fun insertCallAlias(callAlias: String, cc: CtClass): String? {
        if (TextUtil.isEmpty(callAlias)) {
            return null
        }
        if (!callAliasMap.containsKey(callAlias)) {
            callAliasMap[callAlias] = cc.name
            return null
        }
        return callAliasMap[callAlias]
    }
}
